### PR TITLE
WFLY-4923 closing of container managed persistence context should guard against background threads that are also using the persistence context

### PR DESCRIPTION
**Pull request by scottmarlow**
I included the commit from https://github.com/wildfly/wildfly/pull/7820 (need access to TransactionListenerRegistry), after pr#7820 is merged, I will update this pr.

### CODE REVIEW

**Comment by scottmarlow** [comment]
The failure looks legitimate, I'll look into it

File: jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java [hunk 1]
```diff
@@ -45,19 +53,29 @@
 
     private static volatile TransactionSynchronizationRegistry transactionSynchronizationRegistry;
     private static volatile TransactionManager transactionManager;
+    private static volatile TransactionListenerRegistry transactionListenerRegistry;
```
Comment (stuartwdouglas) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java]: All these statics kinda scare me. Why do we need these, as opposed to using services?

Comment (scottmarlow) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java]: We need to reference the statics from non-service code (e.g. container managed persistence context).  We sort of guard the statics with the JPAService but that could be improved (the JPAService injector.uninject methods should probably null the statics out).  Beyond that, any other suggestions for improvement here?

Comment (stuartwdouglas) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java]: Why is the container managed persistence context not managed by a service? I thought it was managed by PersistenceUnitServiceImpl?

Comment (scottmarlow) [inline jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java]: The deployed persistence unit is represented by a (PersistenceUnitServiceImpl) service but I didn't create a service for each container managed persistence context.  I suppose that could also be done with the extended persistence context as well.  The down side of using services for the container managed persistence contexts is that the number of them could be high, which at the least, could consume more constants for representing the service name (and whatever else overhead comes with those changes). Still, its an interesting idea, thanks!

**Comment by scottmarlow** [comment]
It is getting too late in the WildFly 10 development cycle to make this change, will close this pull request and update the jira target to WildFly 11.

**Pull request closed without merging**